src/Evolution/Systems/GeneralizedHarmonic/DampedHarmonicGaugeQuantities.cpp

```diff
@@ -0,0 +1,928 @@
+ * \f]
+ *
+ * contain the temporal and spatial roll-on functions.
+ * */
+namespace {
+template <size_t SpatialDim, typename Frame, typename DataType>
+struct HBuffer;
+
+template <size_t SpatialDim, typename Frame>
+struct HBuffer<SpatialDim, Frame, DataVector> {
+  private:
+    // We make one giant allocation so that we don't thrash the heap.
+    Variables<tmpl::list<::Tags::Tempa<0, SpatialDim, Frame>,
+                        ::Tags::TempScalar<1>,
+                        ::Tags::TempScalar<2>,
+                        ::Tags::TempScalar<3>,
+                        ::Tags::TempScalar<4>,
+                        ::Tags::TempScalar<5>,
+                        ::Tags::TempScalar<6>,
+                        ::Tags::TempScalar<7>,
+                        ::Tags::TempScalar<8>>> buffer_;
+
+  public:
+    explicit HBuffer(const size_t size) noexcept : buffer_(size),
+      spacetime_unit_normal_one_form(get<::Tags::Tempa<0, SpatialDim, Frame>>
+        (buffer_)),
+      log_fac_1(get<::Tags::TempScalar<1>>(buffer_)),
+      log_fac_2(get<::Tags::TempScalar<2>>(buffer_)),
+      weight(get<::Tags::TempScalar<3>>(buffer_)),
+      mu_L1(get<::Tags::TempScalar<4>>(buffer_)),
+      mu_S(get<::Tags::TempScalar<5>>(buffer_)),
+      mu_L2(get<::Tags::TempScalar<6>>(buffer_)),
+      h_prefac1(get<::Tags::TempScalar<7>>(buffer_)),
+      h_prefac2(get<::Tags::TempScalar<8>>(buffer_)) {}
+
+    tnsr::a<DataVector, SpatialDim, Frame>& spacetime_unit_normal_one_form;
+    Scalar<DataVector>& log_fac_1;
+    Scalar<DataVector>& log_fac_2;
+    Scalar<DataVector>& weight;
+    Scalar<DataVector>& mu_L1;
+    Scalar<DataVector>& mu_S;
+    Scalar<DataVector>& mu_L2;
+    Scalar<DataVector>& h_prefac1;
+    Scalar<DataVector>& h_prefac2;
+};
+}  // namespace
+
+template <size_t SpatialDim, typename Frame>
+void compute_damped_harmonic_h(
+    const gsl::not_null<typename GaugeH<SpatialDim, Frame>::type*> gauge_H,
+    const typename InitialGaugeH<SpatialDim, Frame>::type& GaugeH_init,
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, SpatialDim, Frame>& shift,
+    const Scalar<DataVector>& sqrt_det_spatial_metric,
+    const tnsr::aa<DataVector, SpatialDim, Frame>& spacetime_metric,
+    const double& time, const tnsr::I<DataVector, SpatialDim, Frame>& coords,
+    const double& amp_coef_HI,
+    const double& amp_coef_L1,
+    const double& amp_coef_L2,
+    const double& amp_coef_S,
+    const int& exp_L1, const int& exp_L2, const int& exp_S,
+    const double& t_start_HI, const double& sigma_t_HI,
+    const double& t_start_L1,  const double& sigma_t_L1,
+    const double& t_start_L2,  const double& sigma_t_L2,
+    const double& t_start_S,  const double& sigma_t_S,
+    const double& r_max, const double& r_width) noexcept {
+  if (get_size(get<0>(*gauge_H)) != get_size(get(lapse))) {
+    typename GaugeH<SpatialDim, Frame>::type gauge_H_(get(lapse));
+    *gauge_H = gauge_H_;
+  }
```

Comment:
> We assume you want `gauge_H` to be a pointer to avoid repeated allocations. You could comment on that in the header file. If you want to make sure its size is correct, better add an assert, since it would be a programming error.

I can imagine that a single element is p-refined and changes its DataVector size. In that case, we still want this function to work even if called with an outdated sized `gauge_H`. However I have decreased the cost of evaluating this `if` by adding a branch predictor `UNLIKELY` before its argument, as we expect the sizes to be the same more often than not.

Reply:
#1319